orttraining/orttraining/training_ops/cuda/math/softmax_grad_impl.cu

```diff
@@ -28,9 +28,112 @@
+    for (int i = 0; i < WARP_BATCH; ++i) {
+      int batch_element_count = (i >= local_batches) ? 0 : element_count;
+      for (int it = 0; it < WARP_ITERATIONS; ++it) {
+        int element_index = local_idx + it * WARP_SIZE;
+        if (element_index < batch_element_count) {
+          grad_reg[i][it] = grad[i * element_count + it * WARP_SIZE];
+          output_reg[i][it] = output[i * element_count + it * WARP_SIZE];
+          grad_output_reg[i][it] = grad_reg[i][it] * output_reg[i][it];
+        } else {
+          grad_reg[i][it] = acc_t(0);
+          output_reg[i][it] = acc_t(0);
+          grad_output_reg[i][it] = acc_t(0);
+        }
+      }
+    }
+
+    acc_t sum[WARP_BATCH];
+    if (!is_log_softmax) {
+      #pragma unroll
+      for (int i = 0; i < WARP_BATCH; ++i) {
+        sum[i] = grad_output_reg[i][0];
+        #pragma unroll
+        for (int it = 1; it < WARP_ITERATIONS; ++it) {
+          sum[i] += grad_output_reg[i][it];
+        }
+      }
+      warp_reduce<acc_t, WARP_BATCH, WARP_SIZE, Add>(sum);
+    }
+    else {
+      #pragma unroll
+      for (int i = 0; i < WARP_BATCH; ++i) {
+        sum[i] = grad_reg[i][0];
+        #pragma unroll
+        for (int it = 1; it < WARP_ITERATIONS; ++it) {
+          sum[i] += grad_reg[i][it];
+        }
+      }
+      warp_reduce<acc_t, WARP_BATCH, WARP_SIZE, Add>(sum);
+    }
+
+  // store result
+  #pragma unroll
+    for (int i = 0; i < WARP_BATCH; ++i) {
+      if (i >= local_batches)
+        break;
+  #pragma unroll
+      for (int it = 0; it < WARP_ITERATIONS; ++it) {
+        int element_index = local_idx + it * WARP_SIZE;
+        if (element_index < element_count) {
+          // compute gradients
+          if (is_log_softmax) {
+            gradInput[i * element_count + it * WARP_SIZE] = (grad_reg[i][it] - std::exp(output_reg[i][it]) * sum[i]);
+          } else {
+            gradInput[i * element_count + it * WARP_SIZE] = (grad_reg[i][it] - sum[i] ) * output_reg[i][it];
+          }
+        }
+      }
+    }
+  }
+
+// The function "softmax_warp_backward1" saves intermediate results in float32 using registers to prevent recomputing, which can be beneficial for small shapes.
+// However, for larger shapes, the usage of a large register resource can lead to low CUDA warp occupancy and poor performance.
+// In contrast, "softmax_warp_backward2" recomputes intermediate results instead of saving them and also saves the inputs in float16 format to further reduce register usage.
+// TODO: If the dimension to do softmax is greater than 2048, saving the input into shared memory can further reduce register usage.
 template <typename input_t, typename output_t, typename acc_t, int log2_elements, bool is_log_softmax>
-__global__ void softmax_warp_backward(output_t* gradInput, const input_t* grad, const input_t* output,
+__global__ void softmax_warp_backward2(output_t* gradInput, const input_t* grad, const input_t* output,
```

Comment:
shall we use a more self-explained name for example softmax_warp_register_efficient_backward?

Reply:
i am using macro to select diff function, if we rename these two functiones with longer name, it will be tedious to repeat them.

![image](https://user-images.githubusercontent.com/43435212/231127703-6fee641f-1df6-49a3-b7f6-600126efe431.png)

Reply:
i have put a detailed comments above the "softmax_warp_backward2" definition

Reply:
I think it is not hard to make the condition neat: 

```
#define LAUNCH_SOFTMAX_WRAP_BACKWARD_ON_CONDITION(kernel_name)                                                 \
  if (is_log_softmax) {                                                                                        \
    kernel_name<T, T, AccT, log2_elements_value, true>                                                         \
        <<<blocks, threads, 0, stream>>>(input_grad, output_grad, softmax_output, element_count, batch_count); \
  } else {                                                                                                     \
    kernel_name<T, T, AccT, log2_elements_value, false>                                                        \
        <<<blocks, threads, 0, stream>>>(input_grad, output_grad, softmax_output, element_count, batch_count); \
  }

#define CASE_LOG2_ELEMENTS(log2_elements_value)                                           \
  case log2_elements_value: {                                                             \
    if (log2_elements_value > 10) {                                                       \
      LAUNCH_SOFTMAX_WRAP_BACKWARD_ON_CONDITION(softmax_warp_register_efficient_backward) \
    } else {                                                                              \
      LAUNCH_SOFTMAX_WRAP_BACKWARD_ON_CONDITION(softmax_warp_backward)                    \
    }                                                                                     \
  } break
CASE_LOG2_ELEMENTS(0);   // 1
CASE_LOG2_ELEMENTS(1);   // 2
CASE_LOG2_ELEMENTS(2);   // 4
CASE_LOG2_ELEMENTS(3);   // 8
CASE_LOG2_ELEMENTS(4);   // 16
CASE_LOG2_ELEMENTS(5);   // 32
CASE_LOG2_ELEMENTS(6);   // 64
CASE_LOG2_ELEMENTS(7);   // 128
CASE_LOG2_ELEMENTS(8);   // 256
CASE_LOG2_ELEMENTS(9);   // 512
CASE_LOG2_ELEMENTS(10);  // 1024
CASE_LOG2_ELEMENTS(11);  // 2048
```